ipcache: Remove synchronous CIDR identity allocation

gandro opened the pull request:
This PR removes the `AllocateCIDRs` and `ReleaseCIDRIdentitiesByCIDR` functions from the IPCache. Their last usage has been removed when the `ToServices` and `ToFQDN` implementations were moved to use the asynchronous API (using
`{Upsert,Remove}Prefixes`).

Ref: #21142

gandro committed:
ipcache: Remove synchronous CIDR identity allocation

This commit removes the `AllocateCIDRs` and
`ReleaseCIDRIdentitiesByCIDR` functions from the IPCache. Their last
usage has been removed when the `ToServices` and `ToFQDN`
implementations were moved to use the asynchronous API (using
`{Upsert,Remove}Prefixes`).

This commit removes the public functions, as well as the unit test
checking the interaction between the synchronous and asynchronous
methods. The other unit tests are changed to simulate `UpsertPrefixes`
instead.  A subsequent commit remove further now-dead code.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
ipcache: Remove `ReleaseSlice` helper

This function is no longer used now that `ReleaseCIDRIdentitiesByCIDR`
was removed in a previous commit.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
ipcache: Remove `asyncPrefixReleaser`

This removes the asynchronous prefix garbage collector. This is now
possible because the `enqueue` function is no longer called. It was
previously called by `ReleaseCIDRIdentitiesByCIDR`, but that function
has been removed. Therefore, all the dequeue logic can also be removed,
as the gc queue will always be empty.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

tklauser commented on pkg/ipcache/cidr.go:
```diff
@@ -18,78 +18,6 @@ import (
-		newlyAllocatedIdentities = map[netip.Prefix]*identity.Identity{}
-	}
-
-	allocateCtx, cancel := context.WithTimeout(context.Background(), option.Config.IPAllocationTimeout)
```
It looks like `option.Config.IPAllocationTimeout` is now unused, so I guess we could deprecate the corresponding `ip-allocation-timeout` option for v1.16 and remove it for v1.17.

gandro replied:
Good catch! I decided to just remove the flag. It would be a no-op in v1.16, I don't think we need to keep a no-op flag around.

tklauser replied:
AFAIR we usually deprecate user-visible config flags for one release to give users a chance to adjust their config. But I can't find a section in the docs specifying this, only the release notes listing deprecated/removed flags, i.e. https://docs.cilium.io/en/stable/operations/upgrade/#deprecated-options

But given that the `ip-allocation-timeout` option is not used in the Helm chart it's probably OK to just remove it.

tklauser reviewed (approved):
:broom: Nice cleanup! One remark about a now unused option that could be deprecated. Feel free to address as a follow-up.

gandro committed:
daemon: Remove `ip-allocation-timeout` flag

This removes the `ip-allocation-timeout` flag from cilium-agent. It was
used by the now removed `AllocateCIDRs` function. As the timeout does
apply to the asynchronous allocation mechanism, it is removed.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

nathanjsweet reviewed (approved):
Nice.
